### ISSUE TITLE
Streamline RPMs and debs

### ISSUE THREAD

**Issue by runcom** [kubernetes-sigs/cri-o#616]
I would love to have RPMs and debs built as part of our CI (shouldn't impact on time). Those packages will then be available for a smoke test for everyone (and for some automation test).

Spec file and the equivalent for deb should also reside on this git repo - we, maintainers and contributors, know everytime if something changed in the installation and should update spec asap.

@mrunalp @sameo wdyt?

**Comment by rhatdan** [comment]
Seems like a good idea to me.

**Comment by rhatdan** [comment]
@runcom @lsm5 Is this still something we need.  I think @baude did something to get packages to build on ever Pull for podman and buildah.

**Comment by lsm5** [comment]
afaik the COPR approach that @baude uses isn't doable for cri-o since we don't have admin privileges on the kubernetes-incubator account.

I can resume the `make rpm` approach with specfile in upstream if @cevich @edsantiago are on board.

**Comment by cevich** [comment]
/me shares his spare 2x4 with @edsantiago

**Comment by edsantiago** [comment]
@runcom do you still find this necessary? If so, do you consider it nice-to-have or pretty-important?

Reason for asking: long ago I felt ok with keeping specfile & source together. Over the years, experience has changed my mind: distro packaging does not belong in the source repo. In November we had a time-consuming and fruitless exercise in which we tried to combine sources and spec: it was a mess. I think further such attempts will also lead to aggravation.

**Comment by lsm5** [comment]
On Tue, Mar 20, 2018 at 02:15:57PM -0700, Ed Santiago wrote:
> @runcom do you still find this necessary? If so, do you consider it nice-to-have or pretty-important?
> 
> Reason for asking: long ago I felt ok with keeping specfile & source together. Over the years, experience has changed my mind: distro packaging does not belong in the source repo. In November we had a time-consuming and fruitless exercise in which we tried to combine sources and spec: it was a mess. I think further such attempts will also lead to aggravation.


While I used to hate specfile_in_upstream too, it does have value in that we
could just run `make rpm` and
report buildability status, kinda like what I had with the -testonly.spec. Deps
usually don't change that often and runtime deps shouldn't affect buildability
anyway. I don't see the need to keep this specfile totally in sync with
dist-git or to even pull from dist-git everytime.
As long as it builds latest commit/PR, it's good enough.

I can have a cron job to build for rawhide nightlies and we can use these
builds to run whatever other tests necessary.

This approach seems quite simple and cheap. Let me know..

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by wking** [comment]
> I can have a cron job to build for rawhide nightlies and we can use these builds to run whatever other tests necessary.

But this approach would work fine regardless of whether the packaging information was stored in this repo or not, right?  You'd just replace [this][1] with a `curl` call to pull the release source from GitHub.

[1]: https://github.com/kubernetes-incubator/cri-o/commit/9bace79e3503ede55258f69c3dfed4c19d2a8f5d#diff-b67911656ef5d18c4ae36cb6741b7965R96

**Comment by lsm5** [comment]
On Tue, Mar 20, 2018 at 09:52:41PM -0700, W. Trevor King wrote:
> > I can have a cron job to build for rawhide nightlies and we can use these builds to run whatever other tests necessary.
> 
> But this approach would work fine regardless of whether the packaging information was stored in this repo or not, right?  You'd just replace [this][1] with a `curl` call to pull the release source from GitHub.

maybe, but I'd like to avoid change commit/fetch commit stuff in dist-git's
specfile for building the latest commit.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by wking** [comment]
> maybe, but I'd like to avoid change commit/fetch commit stuff in dist-git's specfile for building the latest commit.

`git clone --depth=1 …` is pretty efficient, but if you want to build non-releases you can also pull [an archive for the latest commit][1] from GitHub.

[1]: https://github.com/kubernetes-incubator/cri-o/archive/master.tar.gz

**Comment by lsm5** [comment]
On Wed, Mar 21, 2018 at 09:34:17AM -0700, W. Trevor King wrote:
> > maybe, but I'd like to avoid change commit/fetch commit stuff in dist-git's specfile for building the latest commit.
> 
> `git clone --depth=1 …` is pretty efficient, but if you want to build non-releases you can also pull [an archive for the latest commit][1] from GitHub.
> 
> [1]: https://github.com/kubernetes-incubator/cri-o/archive/master.tar.gz
> 

I don't quite see what you mean.
I thought the point was to test a PR to check if it builds before merge. I
guess I'll leave it to you to send a patch.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by wking** [comment]
> I thought the point was to test a PR to check if it builds before merge.

You're right, we wouldn't be doing that if the RPM spec and other packaging lived in external repositories.  Do we need that though?  We already have local CI to test that things build here.  And you'd have [your daily cron jobs][1] to test the packaging and see if things built on rawhide.  Maybe the need to keep everything together would be easier for me to understand if you could give an example of the kind of packaging failure you're trying to detect in CI?

[1]: https://github.com/kubernetes-incubator/cri-o/issues/616#issuecomment-374809465

**Comment by lsm5** [comment]
On Wed, Mar 21, 2018 at 06:31:29PM +0000, W. Trevor King wrote:
> > I thought the point was to test a PR to check if it builds before merge.
> 
> You're right, we wouldn't be doing that if the RPM spec and other packaging lived in external repositories.  Do we need that though?  We already have local CI to test that things build here.  And you'd have [your daily cron jobs][1] to test the packaging and see if things built on rawhide.  Maybe the need to keep everything together would be easier for me to understand if you could give an example of the kind of packaging failure you're trying to detect in CI?

I don't know if things have changed in cri-o, but I vaguely remember rpms
having build failures in docker sometime in the past. If you're certain those
don't occur anymore, then we don't need this.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by wking** [comment]
> I don't know if things have changed in cri-o, but I vaguely remember rpms having build failures in docker sometime in the past. If you're certain those don't occur anymore, then we don't need this.

Since #43 building the Docker container and running integration tests inside it have been part of Travis testing.  For a recent example, see [here][1] and [here][2].

[1]: https://travis-ci.org/kubernetes-incubator/cri-o/jobs/356427034#L641
[2]: https://travis-ci.org/kubernetes-incubator/cri-o/jobs/356427034#L3190

**Comment by cevich** [comment]
I'm not saying wrong/right here, just a summary of the historical discussions on this topic.  It's totally fine if we rely on the build systems to catch problems (and we should do that also).  My understanding is/was (maybe wrong), we'd like to CI-test the packaged PR artifact because:

*  Because humans are involved, it's possible/likely that ``make all`` from the source-repo produces substantially different binaries than the build system (even against the same source).  Remember, there are multiple release-branches here, and they build against some specific external dependencies.  Running tests for a PR, against a package quickly verifies the, "woops, I forgot to update the spec." class of problems.

* Verifying an artifact from the build-system is typically a costly operation (time-wise), and it can hinder debugging since it's the composite of multiple (already merged) changes.  Testing for the 'likely' outcome based on a built-package, from an individual PR, should help catch some problems closer to their introduction - it's a much tighter feedback-loop.

* End-users will most of the time, install/use cri-o from it's packaged form, not building from source.  So it's beneficial to test cri-o the same way.  It keeps CI-reality closer to real-reality :smile: 

* As above, consumers of the output from `make <package-flavor>`, don't care where/how the composition details came from.  One of those 'consumers' can easily be the PR CI-tests.  So in all cases, wherever the spec is stored won't matter as long as `make` does the expected thing.

**Comment by wking** [comment]
> Running tests for a PR, against a package quickly verifies the, "woops, I forgot to update the spec." class of problems.

With the RPM spec in a separate repo (or even a separate branch in this repo), you could still have CI testing to pick up these sorts of issues.  Spec-branch CI looking for “woops, I forgot to update the spec” is a perfect match.  The issue would be CRI-O-source bugs that only show up in spec CI testing, and I think *those* would be better served by improving the package-agnostic CI tests.

As far as different builds are concerned, that's what Travis' build matrix is for.  We can expand that to cover with and without seccomp, etc. as needed; but I don't see why we'd need to include an RPM spec or other packaging to do that.

> Verifying an artifact from the build-system is typically a costly operation (time-wise)…

This issue is about CI, right?  And it's really about “which branch/repo should these CI tests live in?”.  I don't see a big cost impact depending on whether the packaging CI tests are grouped into the main CRI-O branches or separated into their own branches/repos.

> End-users will most of the time, install/use cri-o from it's packaged form, not building from source. So it's beneficial to test cri-o the same way.

I agree that the packaging (wherever it lives) should use the k8s end-to-end and CRI-O `localintegration` tests as part of its CI.  With solid CRI-O testing, the only failures that should turn up are “hey, we didn't compile/install that correctly”.  If it turns up a CRI-O-source error, the CRI-O tests should be improved.  For an example of this in practice, #1459 is a fix for a CRI-O-source error I turned up while working on the [Gentoo packaging][1].

My main concern with in-master packaging is that the core maintainer set will be unfamiliar with all packaging systems.  Packaging for a given system is a separate skill-set from developing the CRI-O core, and by keeping the packaging separate it's easier for folks to focus on the task at hand.  Core maintainers aren't distracted by packaging noise, and packager's aren't distracted by dev noise.

Where a single person is both a core maintainer and a packager, having separate branches/repos is slightly more work when there's a cross-cutting change (e.g. a new or bumped unvendored dependency).  Do we expect lots of those?  Looking [here][2], I don't see much cross-cutting.  Is there a better place to look to see that sort of thing?

[1]: https://github.com/D3fy/defiance-overlay/pull/11
[2]: https://github.com/lsm5/cri-o/commits/xenial/debian

**Comment by cevich** [comment]
> you could still have CI testing to pick up these sorts of issues.

True, though in many cases the "you" in that sentence is somebody who's juggling a TON of automation/CI priorities or a developer who likes to stay away from CI :fearful:  Hence the 'two birds, one stone' aspect to some of the bullet-points.  But I'm reserving my right to be dead-wrong about all of it :grinning: 

> that's what Travis' build matrix is for

There's oodles of dislike (out of unrelated frustrations) for Travis among this crowd, however what you say could also apply to the origin-ci rigging we've got.  It's also (AFAIK) not strictly/tightly bound to being only RHEL or Fedora.

>> Verifying an artifact from the build-system is typically a costly operation (time-wise)…
> This issue is about CI, right?

Oops, no, that's general bug debugging, where said bug could have been caught at the PR testing stage (and/or the packaging CI stage, if we go that way).   Should have been more clear.

> I agree that the packaging (wherever it lives) should use the k8s end-to-end and CRI-O localintegration tests as part of its CI. 

We have a solution creeping forward in Fedora-land that will solve that.  We may have one in RHEL as well (just heard about it last week).  So the two-birds-one-stone thing appears to be turning into less of a benefit and more of an unnecessary complication.

> My main concern with in-master packaging is that the core maintainer set will be unfamiliar with all packaging systems.

Aye, and we all wear many hats, but I think most here agree with that.  I think you're right about packaging + code synchronization "whoopsies" being better caught in separate packaging CI.   Someone called out a related, specific-whoopsie instance from the past, so I was reiterating that.   Again, the pre-history (of two weeks ago), was that we didn't have a good solution for RHEL packaging CI (nor a $CI_GUY, to set it up).  If $CI_GUY can kill that goose, that problem goes away, and I can relax back to being simply whiny and overworked :laughing: 

-----

To summarize, you think:

*  The Makefile shouldn't be in the business of building packages (I may be reading between some lines)
*  The details about $DISTRO packaging don't belong intermixed in among it's subject-code.
*  We should rely on $DISTRO build/packaging CI to do confirm absence of any version/dependency "whoopsies".

**Comment by wking** [comment]
On Wed, Mar 21, 2018 at 09:55:52PM +0000, Chris Evich wrote:
> > you could still have CI testing to pick up these sorts of issues.
>
> True, though in many cases the "you" in that sentence is somebody
> who's juggling a TON of automation/CI priorities or a developer who
> likes to stay away from CI :fearful: Hence the 'two birds, one
> stone' aspect to some of the bullet-points.

I'm not arguing for anything to be outside of CI.  I'm just arguing
for master to be core code covered by core CI, with packaging
somewhere else with its own CI.

> It's also (AFAIK) not strictly/tightly bound to being only RHEL or
> Fedora.

I don't really care which CI system we use.  But part of my desire to
keep packaging external is that there are *lots* of package managers,
and if we set a packaging-in-master precedent with some of them,
you're asking for “do we have the energy/knowledge to maintain
$other_system packaging in master?”  discussions for each of them.
With out-of-master packaging, whoever has the energy/knowledge can
setup their packaging repo wherever they like without bothering
master.

> Again, the pre-history (of two weeks ago), was that we didn't have a
> good solution for RHEL packaging CI (nor a $CI_GUY, to set it up).

That sounds orthogonal to whether packaging gets its own branch.  If
you can write CI tests for packaging in master, you should be able to
write them in another branch/repo, no?  And conversly, if you can't
write CI tests in another branch/repo, how would putting them in
master here help you?

> To summarize, you think:
>
> * The Makefile shouldn't be in the business of building packages (I
>   may be reading between some lines)

The ‘master’ makefile shouldn't be.  I'm fine with an rmp (and
rpm-1.9, etc.) branch in this repo which adds whatever it likes.  And
I'm also fine if that packaging code lives in another repo.  And I
don't personally mind if folks want to drop packaging code into the
release-1.9, etc. branches, since I'm not worried about maintenance
branches at the moment ;).
 
> * The details about $DISTRO packaging don't belong intermixed in
>   among it's subject-code.

Right.  Packagers are often entirely separate, and the subject code
should be buildable without RPM, dpkg, etc.  Even if we decide to
develop source and packaging in the same branch, we should try not to
mix them together much, because at some point we're likely going to
have to tease them apart again.  But I don't think anyone's arguing
for binding these too tightly, even if they are proponents of keeping
both in master.  Having both sorts of entries in the root Makefile
wouldn't be my favorite from a maintenance perspective, but it
wouldn't be hard to untangle later either.

> * We should rely on $DISTRO build/packaging CI to do confirm absence
>   of any version/dependency "whoopsies".

Somewhat.  We are currently very specific about some dependencies
(e.g. [1]) and less-so about others (e.g. [2]).  Things like [3] are
workarounds for those imprecise dependency declarations (I think,
otherwise I don't understand the difference between btrfs_noversion
and exclude_graphdriver_btrfs).  And I'm in favor of having CI for
spec dependencies (e.g. [4]) as well as having CI for the more
packaging-agnostic Dockerfile.  I think that the CI exercising the RPM
spec should live in the dev branch for the RPM spec, but I don't think
there's any disagreement there.  And the CI for the Dockerfile should
live in master (like it does now) where that Dockerfile is developed.

I just don't see a benefit to having that RPM-spec branch be the CRI-O
master branch, other than having only one branch to bump for
cross-cutting changes.  And I expect cross-cutting changes to be rare.

[1]: https://github.com/kubernetes-incubator/cri-o/blob/0fb585abb483d8b8798a9ab27f8cb73bb1c0d7ca/Dockerfile#L59-L69
[2]: https://github.com/kubernetes-incubator/cri-o/blob/0fb585abb483d8b8798a9ab27f8cb73bb1c0d7ca/Dockerfile#L33
[3]: https://github.com/kubernetes-incubator/cri-o/blob/0fb585abb483d8b8798a9ab27f8cb73bb1c0d7ca/hack/btrfs_tag.sh#L3-L6
[4]: https://github.com/kubernetes-incubator/cri-o/pull/1161/files#diff-e76605a8d55196e7d3ae8d9df3519ce3L23

**Comment by cevich** [comment]
Thanks for spelling it all out.  I'm pretty sure I'm on the same 2x4 as you also now. If it were simple, it wouldn't be called "work" right? :grinning:

**Comment by lsm5** [comment]
On Wed, Mar 21, 2018 at 11:09:43PM +0000, Chris Evich wrote:
> Thanks for spelling it all out.  I'm pretty sure I'm on the same 2x4 as you also now. If it were simple, it wouldn't be called "work" right? :grinning: 

tl;dr, but anyway, I plan to rebuild rawhide nightlies (can do more often if
needed) using master for cri-o and others. Please lmk if I shouldn't.

For xenial debs, I'm already building release-1.[8.9.10] branches and will set
them up for nightlies starting tonight.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by cevich** [comment]
@lsm5 instead of crond, it's possible to convince Travis to do periodic builds.  Then it's not dependant on your specific crond-box.  Just an idea.

**Comment by lsm5** [comment]
On Thu, Mar 22, 2018 at 07:32:19AM -0700, Chris Evich wrote:
> @lsm5 instead of crond, it's possible to convince Travis to do periodic builds.  Then it's not dependant on your specific crond-box.  Just an idea.

well travis uses an older ubuntu, so that's a deal breaker for me.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by cevich** [comment]
@lsm5 there's a way around that, I did a demo this week. :smile: 

Check out: https://github.com/containers/storage/blob/master/.travis.yml and the scripts it uses.

**Comment by lsm5** [comment]
On Thu, Mar 22, 2018 at 02:48:11PM +0000, Chris Evich wrote:
> @lsm5 there's a way around that, I did a demo this week. :smile: 
> 
> Check out: https://github.com/containers/storage/blob/master/.travis.yml and the scripts it uses.

Btw, do you know if this'll allow for unmanned krb auths? For example, fedpkg
needs me to be kinit'd before I can do anything.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by wking** [comment]
> @lsm5 there's a way around that...

Is the trick using the initial Travis environment as a shim to launch [a container with your real CI environment][1]?  That sounds like a good idea to me :). 

[1]: https://github.com/containers/storage/commit/02a59b03989284f02820a56e5c51f98db8306843#diff-85c6256f22fb9c9706deb8bf4f6e3d67R11

**Comment by wking** [comment]
> Is the trick using the initial Travis environment as a shim to launch a container with your real CI environment?

Looks like Travis has an open issue for skipping the shim and directly launching from a user-specified image: travis-ci/travis-ci#7726.  And [GitLab Runner already supports this][1].

[1]: https://docs.gitlab.com/ce/ci/docker/using_docker_images.html#define-image-and-services-from-gitlab-ci-yml

**Comment by cevich** [comment]
> Btw, do you know if this'll allow for unmanned krb auths?

This can be set up through the Travis CI project-settings page.  Essentially you can specify "secret" environment variables who's values will be hidden...unless somebody (with access to .travis.yml) exposes them :face_with_head_bandage: So, it's "possible", though I'd still suggest using a "bot" account with limited privileges, instead of your real one.

GitLab (public and private/personal instances) can also do this.  The mechanism is basically identical to that of Travis, just a different color :smiley_cat:

**Comment by wking** [comment]
With #957 just closed, do we want to close this one as well?

**Comment by cevich** [comment]
@wking soon :grinning: I'm helped @lsm5 get squared away with build + tests for the ubuntu packages, using branches on a fork (for now).  The Fedora RPM-based testing is still quite a bit airborne, but for orthogonal reasons.  The RHEL package-testing part (I think) can be made single-prong - thanks to your dialog above - and hence closure of #957.  It's possible we can solve both the Fedora and RHEL parts with similar approaches - the gang and I will talk this over later today.

**Comment by rhatdan** [comment]
@cevich @wking Is this still something tht is important or can I close?
A year since any activity?

**Comment by wking** [comment]
I [still][1] think we can close this.  If there are still [outstanding packaging issues][2], they can be tracked wherever that external packaging lives (or where it will live), without needing this issue in the upstream repository.

[1]: https://github.com/kubernetes-sigs/cri-o/issues/616#issuecomment-376690771
[2]: https://github.com/kubernetes-sigs/cri-o/issues/616#issuecomment-376913773

**Comment by cevich** [comment]
ya close...oh..it's done :smile:
